### ISSUE TITLE
Detect violations of @final if subclass method does not use @overrides

### ISSUE THREAD

**Issue by eeriksp** [mkorpela/overrides#21]
If a method is marked as `@final`, an error should be risen if subclass tries to override it.
Currently it only works if the subclass's method is decorated with `@overrides`, if this decorator is omitted, no error is shown. (I have not tested it, but it seems to be so from looking at the code.)

I suggest to create a class `StrictParent`, from which other classes using the decorators could inherit from. `StrictParent` would have a metaclass, so it could detect the violations of `@foo` even if the subclass's method does not use `@overrides`.

This solution could be used to solve #12 as well.

Since one might need to use the `@final` and `@overrides` features in an ABC class, `StrictParent`'s metaclass should inherit from `ABCMeta` in order to prevent metaclass conflicts.

**Comment by eeriksp** [comment]
This is my first draft: it is not optimized and I have not checked comparability with older Python versions, but it works in Python 3.7

```py
from abc import ABCMeta

class StrictParentMeta(ABCMeta):
    def __new__(mcls, name, bases, namespace, **kwargs):
        cls = super().__new__(mcls, name, bases, namespace, **kwargs)
        for name, value in namespace.items():
            # Actually checking the direct parent should be enough,
            # otherwise the error would have emerged during the parent class checking
            for base in bases:
                base_class_method = getattr(base, name, False)
                if not base_class_method:
                    continue
                # `__finalized__` is added by `@final` decorator
                assert not getattr(base_class_method, '__finalized__', False), \
                    f'Method `{base_class_method}` is finalized in `{base}`, it cannot be overridden'
        return cls

class StrictParent(metaclass=StrictParentMeta):
    "Use this as the parent class for your custom classes"
    pass
```

**Comment by mkorpela** [comment]
This is awesome metaclass!

**Comment by mkorpela** [comment]
I've implemented WIP version of this now in https://github.com/mkorpela/overrides/commit/76b0933f1b5df888b36ca43d92238dd7546a1f6a
I also added enforcement of @overrides decorator when overriding.

**Comment by mkorpela** [comment]
Released in version 2.4. Thank you for this awesomeness!
